Very detailed step-by-step on how to add a dataset (#904)
huggingface/datasets: ADD_NEW_DATASET.md

You are now ready to start the process of adding the dataset. We will basically create the following files:

	- a dataset script, e.g. `squad.py` which will contain the code to download and pre-process the dataset
	- a dataset cart, `README.md` which will contain general informations and tags related to the dataset

And automatically create the following additional meta-data files (using the CLI tools):

	- a metadata file `dataset_configs.json` which will contain checksums and informations about the dataset to garantee that the loading went fine.
	- a dummy-data file `dummy_data.zip` which will contain small example original files to garantee that the loading script is working well in the future.

### Second step: start writing some code

Let's get started.

1. Create a new branch to hold your development changes with the name of your dataset:

	```bash
	git checkout -b a-descriptive-name-for-my-changes
	```

	**do not** work on the `master` branch.

> 2. Create your dataset folder under `datasets/<your_dataset_name>` and create your dataset script under `datasets/<your_dataset_name>/<your_dataset_name>.py`. You can check out other dataset scripts under `datasets` for some inspiration. Note on naming: the dataset class should be camel case, while the dataset name is its snake case equivalent (ex: `class BookCorpus(datasets.GeneratorBasedBuilder)` for the dataset `book_corpus`).

Reviewer: we can add the list of datasets of reference here I think

3. **Make sure you run all of the following commands from the root of your `datasets` git clone.** To check that your dataset works correctly and to create its `dataset_infos.json` file run the command:

	```bash
	python datasets-cli test datasets/<your-dataset-folder> --save_infos --all_configs
	```

> 4. If the command was succesful, you should now create some dummy data. Use the following command to get in-detail instructions on how to create the dummy data:

Reviewer: Maybe mention the new tool ?

Follow these steps in case the dummy data test keeps failing:

- Verify that all filenames are spelled correctly. Rerun the command
	```bash
	python datasets-cli dummy_data datasets/<your-dataset-folder>
	```
	and make sure you follow the exact instructions provided by the command of step 5).

- Your datascript might require a difficult dummy data structure. In this case make sure you fully understand the data folder logit created by the function `_split_generators(...)` and expected by the function `_generate_examples(...)` of your dataset script. Also take a look at `tests/README.md` which lists different possible cases of how the dummy data should be created.

- If the dummy data tests still fail, open a PR in the repo anyways and make a remark in the description that you need help creating the dummy data.

6. Format your code. Run black and isort so that your newly added files look nice with the following command:

	```bash
	make style
	```

> 

Reviewer: add a step to create the dataset card ?

		- The dataset covers or group several sub-datasets or domains that the users may want to access independantly and/or
		- The dataset comprise several sub-part with different features/organizations of the data (e.g. two types of CSV files with different types of columns). Inside a configuration of a dataset, all the data should have the same format (columns) but the columns can change accross configurations.

	* **Splits** are a more fine grained division than configurations. They allow you, inside a configuration of the dataset, to split the data in typically train/validation/test splits. All the splits inside a configuration should have the same columns/features and splits are thus defined for each specific configurations of there are several.

You are now ready to start the process of adding the dataset. We will basically create the following files:

	- a dataset script, e.g. `squad.py` which will contain the code to download and pre-process the dataset
	- a dataset cart, `README.md` which will contain general informations and tags related to the dataset

And automatically create the following additional meta-data files (using the CLI tools):

> 	- a metadata file `dataset_configs.json` which will contain checksums and informations about the dataset to garantee that the loading went fine.

Reviewer: ```suggestion
	- a metadata file `dataset_infos.json` which will contain checksums and informations about the dataset to garantee that the loading went fine.

```

	- a metadata file `dataset_configs.json` which will contain checksums and informations about the dataset to garantee that the loading went fine.
	- a dummy-data file `dummy_data.zip` which will contain small example original files to garantee that the loading script is working well in the future.

4. Create a new branch to hold your development changes with the name of your dataset:

	```bash
	git checkout -b a-descriptive-name-for-my-changes
	```

	**do not** work on the `master` branch.

5. Create your dataset folder under `datasets/<your_dataset_name>` and create a dataset script under `datasets/<your_dataset_name>/<your_dataset_name>.py`.

### Write some code

Now let's get coding :-)

The dataset script is the main entry point to load and process the data.

There is a detailed explanation on how the library and scripts are organized [here](https://huggingface.co/docs/datasets/add_dataset.html).

Note on naming: the dataset class should be camel case, while the dataset short_name is its snake case equivalent (ex: `class BookCorpus(datasets.GeneratorBasedBuilder)` for the dataset `book_corpus`).

> To add a nwew dataset, you can start by copying one of the datasets of reference listed below. The main criteria for choosing to start from one or the other reference dataset is the format of the data files (JSON/CSV/TSV/text) and whether you need several configurations (see above explanations on configurations). Feel free to reuse parts of the code and adapt them to your case:

Reviewer: ```suggestion
To add a new dataset, you can start by copying one of the datasets of reference listed below. The main criteria for choosing to start from one or the other reference dataset is the format of the data files (JSON/CSV/TSV/text) and whether you need several configurations (see above explanations on configurations). Feel free to reuse parts of the code and adapt them to your case:

```

This let you use `print()` statements as well as seeing directly errors and the final dataset format.

**Important rules to follow**:

- try to give access to all the data, columns, features and information in the dataset. If the dataset contains various sub-parts with differing formats, create several configurations to give access to all of them.
- datasets in the `datasets` library are typed. Take some time to carefully think about the `features` (see an introduction [here](https://huggingface.co/docs/datasets/exploring.html#features-and-columns) and the full list of possible features [here](https://huggingface.co/docs/datasets/features.html))
- if some of you dataset features are in a fixed set of classes (e.g. labels), you should use a `ClassLabel` feature. 

### Automatically add code metadata

Now that your dataset script runs and create a dataset with the format you expected.

1. **Make sure you run all of the following commands from the root of your `datasets` git clone.** To check that your dataset works correctly and to create its `dataset_infos.json` file run the command:

	```bash
	python datasets-cli test datasets/<your-dataset-folder> --save_infos --all_configs
	```

> 	This first command should create a `dataset_configs.JSON` file in your dataset folder.

Reviewer: ```suggestion
	This first command should create a `datasets_infos.json` file in your dataset folder.

```
